lib/galaxy/workflow/modules.py

```diff
@@ -1145,6 +1145,8 @@ def callback( input, value, prefixed_name, prefixed_label ):
                         if isinstance( input, DataToolParameter ):
                             # Pull out dataset instance from element.
                             replacement = iteration_elements[ prefixed_name ].dataset_instance
```

Comment:
Have you tried just modifying this line from...

```
replacement = iteration_elements[ prefixed_name ].dataset_instance
```

to

```
replacement = iteration_elements[ prefixed_name ]
```

I'm not sure if there would be a problem elsewhere in the workflow code, but this would feel more like what happens for tools.

If you try that and it doesn't work - this is probably fine for now, but can you add a comment with a TODO comment to populate the the wrappr with the element identifier as an argument and not by modifying the HDA.